### PR TITLE
playground: make jsoo_main compile

### PR DESCRIPTION
**Pull request by jchavarri**
A humble start, just make `Jsoo_main` compile in a jsoo executable. As this part of the code is never exercised in ci, it had rotten quite a bit.

Now, it can be loaded from node, but instantly triggers a cryptic runtime error the moment you try to compile anything.

Repro:
```
opam install js_of_ocaml
dune build jscomp/main/jsoo_main.bc.js
node
Welcome to Node.js v17.8.0.
Type ".help" for more information.
> require('./_build/default/jscomp/main/jsoo_main.bc.js');
{}
> ocaml
{ compile: [Function (anonymous)], version: '0.3.2' }
> ocaml.compile("let t = 1");
Fatal error: exception Melange_compiler_libs.Env.Error(_)
```

### TODO

Not sure if all should be part of this PR, or some parts can be deferred.

- [x] Don't hardcode `BS_BROWSER`: not sure how to set env var only for a dune executable
- [ ] Investigate warning while compiling: `Warning: integer overflow: integer 0x100000000 (4294967296) truncated to 0x0 (0); the generated code might be incorrect.`
- [x] Compilation errors: properly handle the case when `error_of_exn` returns `Some`
- [x] Add the compilation of `Jsoo_main` to ci to prevent future rot
- [x] Investigate why building `dune build --profile=prod jscomp/main/jsoo_main.bc.js` crashes (maybe related to `BS_BROWSER`)
- [x] Make both `melange.ppx` and the "internal" ppx apply to the input code

### CODE REVIEW

**Comment by jchavarri** [comment]
Manage to print the internals of the cryptic `Fatal error: exception Melange_compiler_libs.Env.Error(_)` error by updating melange compiler libs:

```
Lookup_error: Unbound_module: Stdlib
```

So it seems for some reason the Melange stdlib is not being "linked" in the resulting `bc.js` file? I tried adding `melange` to the list of libraries in the executable here:

https://github.com/melange-re/melange/blob/6944289695ace23995dfd18542cf0392c7976a24/jscomp/main/dune#L45

But still get the same error.

**Comment by hhugo** [comment]
> Managed to print the internals of the cryptic `Fatal error: exception Melange_compiler_libs.Env.Error(_)` error by updating melange compiler libs:
> 
> ```
> Lookup_error: Unbound_module: Stdlib
> ```
> 
> So it seems for some reason the Melange stdlib is not being "linked" in the resulting `bc.js` file? I tried adding `melange` to the list of libraries in the executable here:
> 
> https://github.com/melange-re/melange/blob/6944289695ace23995dfd18542cf0392c7976a24/jscomp/main/dune#L45
> 
> But still get the same error.

I don't think this mean that the stdlib was not linked in, but rather that the CMI of the stdlib was not found by the type checker

**Comment by hhugo** [comment]
Look at `Persistent_env.Persistent_signature.load := Bs_cmi_load.load_cmi`. it seems to be trying to load cmis using filesystem calls

**Comment by jchavarri** [comment]
@hhugo thanks for the pointers! I started remembering now that I have to set up a process where the cmis and cmjs for the libraries used are placed somewhere, so they can be loaded in the static "pseudo" file system that jsoo uses at runtime.

I found the relevant part of the process in rescript repo, where `js_of_ocaml build-fs` is called [here](https://github.com/rescript-lang/rescript-compiler/blob/c646742680031214ab41bdd0472bac52baa36a8c/packages/playground-bundling/scripts/generate_cmijs.js#L77
).

I understand that for the pseudo file system to work, besides preparing the cm*s like mentioned above, the jsoo build of melange should ignore the `Persistent_signature.load` mutation right? Something like this in `Bs_conditional_initial.setup_env`:

```
  #ifdef BS_BROWSER
  #else
    Persistent_env.Persistent_signature.load := Bs_cmi_load.load_cmi;
  #endif
```

**Comment by jchavarri** [comment]
I got over the Stdlib errors by loading the `Stdlib` cmis in the pseudo file system 🎉  In the end, I didn't have to use `ifdef BS_BROWSER` to avoid the `Persistent_env.Persistent_signature.load` override 🤔 

There is now some error the second time the compiler is invoked that I will investigate next:

```
$ make dev # build melange stdlib cmijs
$ dune build jscomp/main/jsoo_main.bc.js # build js version of the compiler
$ find _build/default/jscomp/stdlib-412/.stdlib.objs/melange -name "*.cmi" -or -name "*.cmj" | xargs js_of_ocaml build-fs -o stdlib.js
$ node
Welcome to Node.js v18.16.0.
Type ".help" for more information.
> require('./_build/default/jscomp/main/jsoo_main.bc.js');
{}
> require('./stdlib.js');
{}
> ocaml.compile("let t = 1");
{
  js_code: '// Generated by Melange\n' +
    "'use strict';\n" +
    '\n' +
    '\n' +
    'var t = 1;\n' +
    '\n' +
    'exports.t = t;\n' +
    '/* No side effect */\n'
}
> ocaml.compile("let t = 1");
{ js_error_msg: 'Invalid_argument("Ext_list.fold_right2")' }
```

**Comment by hhugo** [comment]
> I got over the Stdlib errors by loading the `Stdlib` cmis in the pseudo file system tada In the end, I didn't have to use `ifdef BS_BROWSER` to avoid the `Persistent_env.Persistent_signature.load` override thinking
> 
> There is now some error the second time the compiler is invoked that I will investigate next:
> 
> ```
> $ make dev # build melange stdlib cmijs
> $ dune build jscomp/main/jsoo_main.bc.js # build js version of the compiler
> $ find _build/default/jscomp/stdlib-412/.stdlib.objs/melange -name "*.cmi" -or -name "*.cmj" | xargs js_of_ocaml build-fs -o stdlib.js
> $ node
> Welcome to Node.js v18.16.0.
> Type ".help" for more information.
> > require('./_build/default/jscomp/main/jsoo_main.bc.js');
> {}
> > require('./stdlib.js');
> {}
> > ocaml.compile("let t = 1");
> {
>   js_code: '// Generated by Melange\n' +
>     "'use strict';\n" +
>     '\n' +
>     '\n' +
>     'var t = 1;\n' +
>     '\n' +
>     'exports.t = t;\n' +
>     '/* No side effect */\n'
> }
> > ocaml.compile("let t = 1");
> { js_error_msg: 'Invalid_argument("Ext_list.fold_right2")' }
> ```

There is probably some `reset` function to call in between compilation.

**Comment by jchavarri** [comment]
I could track it down to [this line](https://github.com/melange-re/melange/blob/2f15294f70193f4f49d8caf21f0cb1d86716ec0e/jscomp/core/lam_coercion.ml#LL89C1-L89C1), where the lambda exports and the "meta exports" differ in length. Noticed there is something missing in the `reset` function, will open a PR in melange-compiler-libs.

File: jscomp/main/jsoo_main.ml [hunk 1]
```diff
@@ -54,52 +56,44 @@ let implementation ~use_super_errors impl str : Js.Unsafe.obj =
...
-  if use_super_errors then (
-    Misc.Color.setup (Some Always);
-    Lazy.force Super_main.setup);
-
   try
-    Js_config.jsx_version := 3;
     (* default *)
     let ast = impl (Lexing.from_string str) in
-    let ast = Ppx_entry.rewrite_implementation ast in
+    let ast = Melange_ppx_lib.Ppx_entry.rewrite_implementation ast in
```
Comment (jchavarri) [inline jscomp/main/jsoo_main.ml:62]: @anmonteiro I am using this line with the intention of being able to process `bs.raw` and such extension nodes. But it doesn't seem to work:

```
$ node
> require('./_build/default/jscomp/main/jsoo_main.bc.js');
{}
> require('./stdlib.js');
{}
> ocaml.compile("[%%bs.raw \"var a = 1\"]");
{
  js_error_msg: "Line 1, 3:\n  Error Uninterpreted extension 'bs.raw'.",
  row: 0,
  column: 3,
  endRow: 0,
  endColumn: 9,
  text: "Uninterpreted extension 'bs.raw'.",
  type: 'error'
}
```

Do you have any ideas what I might be missing?

Comment (jchavarri) [reply jscomp/main/jsoo_main.ml:62]: Oh, this is the "other" part of the ppx, that processes the attributes!

```
> ocaml.compile(`type person\nexternal john : person = "john" [@@bs.module "MySchool"]`);
{
  js_code: '// Generated by Melange\n' +
    "/* This output is empty. Its source's type definitions, externals and/or unused code got optimized away. */\n"
}
```

I need to somehow plug in the actual `melange.ppx` that is using the ppxlib driver.

Comment (jchavarri) [reply jscomp/main/jsoo_main.ml:62]: Fixed in https://github.com/melange-re/melange/pull/538/commits/08be8255e8fc4f8955d4c501f3d1f2b04f1aa2fb.

**Comment by jchavarri** [comment]
I think the PR is in a state where it could be reviewed, before adding more functionality to the playground build, like Reason syntax or custom libraries cmi/js.

The only item from the todo list in the description that I couldn't find a fix for are the warnings. I saw there [a line in the melange runtime lib](https://github.com/melange-re/melange/blob/b23d4f44ac28ce45eb82cb7527eceda363461e32/jscomp/runtime/caml_int64.ml#L299) that has a value `0x100000000` but I tried updating it and jsoo would still show the same warning. It doesn't seem to me that the jsoo playground depends on the melange runtime at all, so I am probably looking in the wrong place.

Otherwise, the following works:

```
$ make playground
$ make dev # build melange stdlib cmijs
$ find _build/default/jscomp/stdlib-412/.stdlib.objs/melange -name "*.cmi" -or -name "*.cmj" | xargs js_of_ocaml build-fs -o stdlib.js
$ node
Welcome to Node.js v18.16.0.
Type ".help" for more information.
> require('./_build/default/jscomp/main/jsoo_main.bc.js');
{}
> require('./stdlib.js');
{}
> ocaml.compile("let t = 1");
{
  js_code: '// Generated by Melange\n' +
    "'use strict';\n" +
    '\n' +
    '\n' +
    'var t = 1;\n' +
    '\n' +
    'exports.t = t;\n' +
    '/* No side effect */\n'
}
> ocaml.compile("[%%bs.raw \"var a = 1\"]");
{
  js_code: '// Generated by Melange\n' +
    "'use strict';\n" +
    '\n' +
    '\n' +
    'var a = 1\n' +
    ';\n' +
    '\n' +
    '/*  Not a pure module */\n'
}
> ocaml.compile(`type person\nexternal john : person = "john" [@@bs.module "MySchool"]`);
{
  js_code: '// Generated by Melange\n' +
    "/* This output is empty. Its source's type definitions, externals and/or unused code got optimized away. */\n"
}

File: jscomp/core/js_implementation.mli [hunk 2]
```diff
@@ -37,3 +37,8 @@ val implementation :
 
 val implementation_cmj : Format.formatter -> string -> unit
 (** [implementation_cmj ppf cmj_file] compiles a cmj to JS *)
+
+module Ppx_entry : sig
+  val rewrite_signature : Parsetree.signature -> Parsetree.signature
```
Comment (anmonteiro) [inline jscomp/core/js_implementation.mli:42]: you probably found out that this doesn't do what you think it does. 

You can revert this change, it doesn't need to be exposed.

File: jscomp/main/dune [hunk 3]
```diff
@@ -6,7 +6,7 @@
  (modules melc melc_cli)
  (preprocess
   (action
-   (run cppo "%{env:CPPO_FLAGS=}" %{input-file}))))
+   (run cppo "%{env:CPPO_FLAGS=}" -D BS_BROWSER %{input-file}))))
```
Comment (anmonteiro) [inline jscomp/main/dune:9]: Are we defining this unconditionally? Doesn't this break our non-browser compiler?

File: jscomp/main/dune [hunk 4]
```diff
@@ -26,3 +26,10 @@
  (package melange)
  (modules ounit_tests_main)
  (libraries ounit2 ounit_tests))
+
+(executable
+ (name jsoo_main)
+ (modules jsoo_main jsoo_common)
+ ; Need to use melange_ppx, not melange_ppx_lib, so Driver.V2.register_transformation runs
```
Comment (anmonteiro) [inline jscomp/main/dune:33]: `melange.ppx-lib` is not a thing anymore after #583, the comment can be updated.

File: jscomp/main/jsoo_main.mli [hunk 5]
```diff
@@ -22,4 +22,5 @@
  * along with this program; if not, write to the Free Software
  * Foundation, Inc., 59 Temple Place - Suite 330, Boston, MA 02111-1307, USA. *)
 
-val make_compiler : string -> (Lexing.lexbuf -> Parsetree.structure) -> unit
+val make_compiler :
+  string -> (Lexing.lexbuf -> Melange_compiler_libs.Parsetree.structure) -> unit
```
Comment (anmonteiro) [inline jscomp/main/jsoo_main.mli:26]: we could `-open Melange_compiler_libs` in flags for `jsoo_main` like we do everywhere else

File: jscomp/main/jsoo_common.ml [hunk 6]
```diff
@@ -32,20 +32,44 @@ module Js = struct
   external to_bytestring : js_string t -> string = "caml_js_to_byte_string"
 end
 
-let mk_js_error (loc : Location.t) (msg : string) =
+let string_of_formatted ~f x =
```
Comment (anmonteiro) [inline jscomp/main/jsoo_common.ml:35]: is this `Format.asprintf`?

File: jscomp/main/jsoo_main.ml [hunk 7]
```diff
@@ -54,52 +56,49 @@ let implementation ~use_super_errors impl str : Js.Unsafe.obj =
...
-    Lazy.force Super_main.setup);
-
   try
-    Js_config.jsx_version := 3;
     (* default *)
     let ast = impl (Lexing.from_string str) in
-    let ast = Ppx_entry.rewrite_implementation ast in
+    let ast = Js_implementation.Ppx_entry.rewrite_implementation ast in
+    let ast =
+      Melange_ppxlib_ast.Of_ppxlib.copy_structure
```
Comment (anmonteiro) [inline jscomp/main/jsoo_main.ml:64]: this changed a bit in #604

**Comment by anmonteiro** [comment]
there's a bit of a conflict with #615, you should be good just moving your files to `bin/` when rebasing.

File: jscomp/core/js_implementation.mli (hunk 2 above)
Comment (jchavarri) [reply jscomp/core/js_implementation.mli]: removed in https://github.com/melange-re/melange/pull/538/commits/abdb6ae192772eb3bbd90c280c055e3977d32fbc.

File: jscomp/main/dune (hunk 3 above)
Comment (jchavarri) [reply jscomp/main/dune]: No, this was a remaining from previous experiments that I forgot to remove. Did in https://github.com/melange-re/melange/pull/538/commits/df0a373a9204b3b4a47a02e99065f939f8f3ebba.

File: jscomp/main/jsoo_main.ml (hunk 7 above)
Comment (jchavarri) [reply jscomp/main/jsoo_main.ml]: I am not sure what is the right way to convert these asts now, using a lot of magic... https://github.com/melange-re/melange/pull/538/commits/529df6a7f99e8e04996370261e1b6a9db546348e

File: jscomp/main/dune (hunk 4 above)
Comment (jchavarri) [reply jscomp/main/dune]: https://github.com/melange-re/melange/pull/538/commits/c4ad80d59c355c0478dc3787ad404386d212c11a

File: jscomp/main/jsoo_main.mli (hunk 5 above)
Comment (jchavarri) [reply jscomp/main/jsoo_main.mli]: https://github.com/melange-re/melange/pull/538/commits/1620cbc9024e36c83775c7f4ba2dc7461d2fddf4

File: jscomp/main/jsoo_common.ml (hunk 6 above)
Comment (jchavarri) [reply jscomp/main/jsoo_common.ml]: https://github.com/melange-re/melange/pull/538/commits/9329c8714320bcf9988f1792c243d7b5af89be8f

**Comment by jchavarri** [comment]
Ah, I also added some test for the playground to avoid having to run all these commands in Node manually every time anything changes. I couldn't find a way to have this test run with all the other tests, as the only way to set an env var for the `jsoo_main.bc.js` executable is by setting it from the profile, which means the new test can't be run with all the others, as profiles are global.

File: bin/jsoo_main.ml [hunk 8]
```diff
@@ -54,52 +58,52 @@ let implementation ~use_super_errors impl str : Js.Unsafe.obj =
...
     let ast = impl (Lexing.from_string str) in
-    let ast = Ppx_entry.rewrite_implementation ast in
+    let ast : Parsetree.structure =
+      let ppxlib_ast : Ppxlib_ast__.Versions.OCaml_414.Ast.Parsetree.structure =
+        Obj.magic (ast : Parsetree.structure)
+      in
+      let converted =
+        Convert.copy_structure (Ppxlib.Driver.map_structure ppxlib_ast)
+      in
+      (Obj.magic converted : Parsetree.structure)
```
Comment (anmonteiro) [inline bin/jsoo_main.ml:71]: I'm pretty sure this is wrong. You probably copied it from `melc.ml`.

This is applying the PPXes and converting the preprocessed AST to the _current switch_ version. Also `Parsetree.structure` isn't what you think it is because you `open`ed `Melange_compiler_libs` globally.

In the playground, we just need this AST version to be the Melange_compiler_libs parsetree, I think? so we might not need to make the copy at all?

Comment (anmonteiro) [reply bin/jsoo_main.ml]: I just pushed 53dc2dceba5759a49db71ad00cb59b091e1ba7b9 fixing this

**Comment by anmonteiro** [comment]
The `Obj.magic` situation is a bit unfortunate, because we version our own Parsetree, which is technically the same as the current ppxlib AST.

I haven't found a way to abstract over these that I'm happy with, so I've been repeating slight variations of the same code (now present in 3 places IIRC). I'll clean it up soon.

**Comment by anmonteiro** [comment]
Let's do this.

File: bin/jsoo_main.ml (hunk 8 above)
Comment (jchavarri) [reply bin/jsoo_main.ml]: > You probably copied it from melc.ml.

Yes.

> In the playground, we just need this AST version to be the Melange_compiler_libs parsetree, I think? so we might not need to make the copy at all?

Just curious, why does ppxlib need all this `copy_structure` calls?
